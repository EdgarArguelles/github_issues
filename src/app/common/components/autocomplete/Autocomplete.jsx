import './Autocomplete.scss';
import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {useFocusRef, useSetFocus} from 'app/hooks/Common';
import ResultItem from './result_item/ResultItem';

const Autocomplete = ({placeholder, initValue, onSelect, renderResult}) => {
    const [search, setSearch] = useState('');
    const [results, setResults] = useState(null);
    const [currentFocus, setCurrentFocus] = useSetFocus(results?.length + 1);
    const searchRef = useFocusRef(currentFocus === 0);
    const handleChange = event => {
        const value = event.target.value;
        setSearch(value);
        setResults(value ? renderResult(value) : null);
    };

    const handleKeyDown = event => {
        if (event.keyCode === 27) {
            event.preventDefault();
            setSearch('');
            setResults(null);
        } else if (event.keyCode === 13 && results?.length) {
            event.preventDefault();
            setCurrentFocus(1);
        }
    };

    const handleSelect = item => {
        setResults(null);
        setCurrentFocus(0);
        if (initValue) {
            setSearch(initValue);
        }
        onSelect(item);
    };

    const getResults = () => {
        if (!results.length) {
            return <div className="empty">Not result</div>;
        }
        return results.map((data, index) => (
            <ResultItem key={`result-item-${index}`} data={data} onSelect={handleSelect}
                        focus={currentFocus === index + 1}/>)
        );
    };

    useEffect(() => {
        if (initValue) {
            setSearch(initValue);
        }
    }, [initValue]);

    return (
        <div className="autocomplete">
            <input type="text" className="search" placeholder={placeholder} ref={searchRef}
                   value={search} onChange={handleChange} onKeyDown={handleKeyDown}/>
            {results && <div className="results">{getResults()}</div>}
        </div>
    );
};

Autocomplete.propTypes = {
    placeholder: PropTypes.string,
    initValue: PropTypes.string,
    onSelect: PropTypes.func.isRequired,
    renderResult: PropTypes.func.isRequired,
};

export default React.memo(Autocomplete);