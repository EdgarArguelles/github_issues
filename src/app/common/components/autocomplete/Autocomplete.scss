.autocomplete {
  .search {
    padding: 15px;
    border-color: #80808026;
    border-width: 3px;
    width: calc(100% - 22px);
  }

  .results {
    position: fixed;
    background-color: white;
    border-style: solid;
    border-width: 1px;
    width: inherit;
    max-height: 500px;
    overflow-y: scroll;

    .empty {
      padding: 15px;
    }
  }
}