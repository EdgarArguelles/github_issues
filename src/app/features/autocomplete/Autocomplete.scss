// variables
@use 'app/Variables';

.autocomplete {
  .search {
    padding: 15px;
    border-color: Variables.$search-border-color;
    border-width: 3px;
    width: calc(100% - 22px);
  }

  .results {
    position: fixed;
    background-color: Variables.$results-background-color;
    border-style: solid;
    border-width: 1px;
    width: inherit;
    max-height: 500px;
    overflow-y: scroll;

    .empty {
      padding: 15px;
    }
  }
}