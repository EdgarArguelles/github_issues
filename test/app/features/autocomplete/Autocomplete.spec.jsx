import sinon from 'sinon';
import renderer, {act} from 'react-test-renderer';
import React from 'react';
import Autocomplete from 'app/features/autocomplete/Autocomplete';

describe('Components -> Autocomplete', () => {
    const onSelectStub = sinon.stub();
    const renderResultStub = sinon.stub();
    let component, wrapper;

    afterEach(() => {
        onSelectStub.reset();
        renderResultStub.reset();
    });

    const mountComponent = props => {
        wrapper = renderer.create(<Autocomplete {...props}/>);
        component = wrapper.root.find(el => el.type.name === 'Autocomplete');
    };

    describe('load', () => {
        it('should have required props', () => {
            mountComponent({
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            expect(component.props.placeholder).toBeUndefined();
            expect(component.props.initValue).toBeUndefined();
            expect(component.props.onSelect).toStrictEqual(onSelectStub);
            expect(component.props.renderResult).toStrictEqual(renderResultStub);
        });

        it('should have all props', () => {
            mountComponent({
                placeholder: 'placeholder test',
                initValue: 'initValue test',
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            expect(component.props.placeholder).toStrictEqual('placeholder test');
            expect(component.props.initValue).toStrictEqual('initValue test');
            expect(component.props.onSelect).toStrictEqual(onSelectStub);
            expect(component.props.renderResult).toStrictEqual(renderResultStub);
        });
    });

    describe('renderResult', () => {
        it('should call not renderResult when null', () => {
            mountComponent({
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            act(() => wrapper.root.findByType('input').props.onChange({target: {value: null}}));

            sinon.assert.callCount(renderResultStub, 0);
        });

        it('should call renderResult when value', () => {
            mountComponent({
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            act(() => wrapper.root.findByType('input').props.onChange({target: {value: 'test'}}));

            sinon.assert.callCount(renderResultStub, 1);
            sinon.assert.calledWithExactly(renderResultStub, 'test');
        });
    });

    describe('snapshot', () => {
        it('should render Autocomplete without props', () => {
            mountComponent({
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            expect(wrapper.toJSON()).toMatchSnapshot();
        });

        it('should render Autocomplete with props', () => {
            mountComponent({
                placeholder: 'placeholder test',
                initValue: 'initValue test',
                onSelect: onSelectStub,
                renderResult: renderResultStub,
            });

            expect(wrapper.toJSON()).toMatchSnapshot();
        });
    });
});