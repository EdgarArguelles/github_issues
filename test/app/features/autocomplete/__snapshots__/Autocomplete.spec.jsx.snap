// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components -> Autocomplete snapshot should render Autocomplete with props 1`] = `
<div
  className="autocomplete"
>
  <input
    className="search"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="placeholder test"
    type="text"
    value=""
  />
</div>
`;

exports[`Components -> Autocomplete snapshot should render Autocomplete without props 1`] = `
<div
  className="autocomplete"
>
  <input
    className="search"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</div>
`;
